### PR TITLE
improve type annotations in 'docutils.io.Input'

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

File: stubs/docutils/docutils/io.pyi [hunk 1]
```diff
@@ -23,6 +23,18 @@ def error_string(err: BaseException) -> str: ...
...
+        source: str | None = None,
+        source_path: str | None = None,
+        encoding: str | None = None,
+        error_handler: str = "strict",
+    ) -> None: ...
+    encoding: str | None
+    error_handler: str
+    source: str | None
+    source_path: str | None
+    successful_encoding: str | None = None
```
Comment (srittau) [inline stubs/docutils/docutils/io.pyi:37]: Nit: Could you move this above the `__init__` method as is customary in typeshed?

File: stubs/docutils/docutils/io.pyi [hunk 2]
```diff
@@ -23,6 +23,18 @@ def error_string(err: BaseException) -> str: ...
...
     def read(self) -> Any: ...
```
Comment (srittau) [inline stubs/docutils/docutils/io.pyi:38]: I've noticed this:

```suggestion
    def read(self) -> str: ...
```

File: stubs/docutils/docutils/io.pyi [hunk 3]
```diff
@@ -23,6 +23,18 @@ def error_string(err: BaseException) -> str: ...
 class Input(TransformSpec):
     component_type: ClassVar[str]
     default_source_path: ClassVar[str | None]
+    def __init__(
+        self,
+        source: str | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/io.pyi:28]: Looking at the sub-classes, source seems to be some kind of file-like type. Using a protocol would be best, although using `IO` would do in a pinch.

**Review by srittau — CHANGES_REQUESTED** [review]

File: stubs/docutils/docutils/io.pyi (hunk 3 above)
Comment (danieleades) [reply stubs/docutils/docutils/io.pyi]: changed to `source: IO[str] | None = None`. Is that what you had in mind?

**Comment by srittau** [comment]
Looking at it a bit more, it seems that `source` is whatever the sub-class needs it to be. In the case of `StringInput` it's a `str`, in the case of `FileInput` it's some kind of file-like object. I see two options: Make `Input` generic over `source` or use `Any` and have the sub-classes override the field with whatever they use as source.

**Comment by danieleades** [comment]
> Looking at it a bit more, it seems that `source` is whatever the sub-class needs it to be. In the case of `StringInput` it's a `str`, in the case of `FileInput` it's some kind of file-like object. I see two options: Make `Input` generic over `source` or use `Any` and have the sub-classes override the field with whatever they use as source.

i prefer the generic approach. Likely to have more impact downstream, but mainly on sphinx, but they're not actively using typeshed stubs yet anyway

(the plan is to transition to typeshed, hence my current flurry of activity. See https://github.com/sphinx-doc/sphinx/issues/12007

**Comment by danieleades** [comment]
> ```diff
> sphinx (https://github.com/sphinx-doc/sphinx)
> - sphinx/transforms/i18n.py:74:20: error: Unexpected keyword argument "source" for "StringInput"  [call-arg]
> + sphinx/transforms/i18n.py:74:39: error: Argument "source" to "StringInput" has incompatible type "str"; expected "Optional[IO[str]]"  [arg-type]
> + sphinx/transforms/i18n.py: note: At top level:
> + sphinx/transforms/i18n.py: note: In member "apply" of class "Locale":
> - note: In function "publish_msgstr":
> - note: "StringInput" defined here
> - sphinx/transforms/i18n.py:74:20: error: Unexpected keyword argument "source_path" for "StringInput"  [call-arg]
> - note: "StringInput" defined here
> ```

don't understand this error:

`+ sphinx/transforms/i18n.py:74:39: error: Argument "source" to "StringInput" has incompatible type "str"; expected "Optional[IO[str]]"  [arg-type]`

the signature for StringInput is now

```python
class StringInput(Input[str]):
    default_source_path: ClassVar[str]
```

hence the type of `source` should be `str | None` (not `IO[str]`)

File: stubs/docutils/docutils/io.pyi [hunk 4]
```diff
@@ -20,10 +20,20 @@ class OutputError(OSError): ...
...
     component_type: ClassVar[str]
     default_source_path: ClassVar[str | None]
-    def read(self) -> Any: ...
+    encoding: str | None
+    error_handler: str
+    source: _S | None
+    source_path: str | None
+    successful_encoding: str | None = None
+    def __init__(
+        self, source: _S | None = None, source_path: str | None = None, encoding: str | None = None, error_handler: str = "strict"
```
[the remark this answers is not in the document]
Comment (danieleades) [reply stubs/docutils/docutils/io.pyi:34]: you might want to consider adding the project here - https://github.com/hauntsaninja/mypy_primer/blob/master/mypy_primer/projects.py

**Review by srittau — APPROVED** [review]

**Pull request merged**